src/tests/Common/XUnitWrapperLibrary/TestSummary.cs

```diff
@@ -95,12 +102,66 @@ public string ToXmlString()
 
             return testResultSb.ToString();
         }
+
+        private string SanitizeOutput(string output)
+        {
+            Dictionary<char, string>? invalidChars = FindInvalidXMLChars(output);
+
+            // All green, we can print the test's output as is.
+            if (invalidChars is null)
+                return output;
+
+            StringBuilder sanitizedOutput = new StringBuilder();
+            foreach (char ch in output)
+            {
+                string? invalidCharHexCode = string.Empty;
```

Comment:
Why not just do a Replace on the string for each of the problematic characters rather than walking the string yourself? You know the string you are replacing with does not contain any invalid characters

Reply:
Although I'm not sure why you don't just walk the steering checking each char is valid and replacing it on the fly rather than building up a dictionary

Reply:
> Why not just do a Replace on the string for each of the problematic characters rather than walking the string yourself? You know the string you are replacing with does not contain any invalid characters

`String.Replace()` always creates a new string so it's inefficient to be creating new ones for each problematic character, especially considering some tests might have large outputs, and we have lots of tests to process.

Reply:
> Although I'm not sure why you don't just walk the string checking each char is valid and replacing it on the fly rather than building up a dictionary first?

I did it to just return the output as is, in case there were no problematic characters, but yeah it might be simpler as you suggest. Will try it right now.

Reply:
> String.Replace() always creates a new string so it's inefficient to be creating new ones for each problematic character

Right, for several replacements you want StringBuilder.Replace.